File: 10258 Contest Scoreboard/solution.java
import java.io.*;
import java.util.*;

public class Main {
	public static void main(String[] args) throws IOException {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StreamTokenizer st = new StreamTokenizer(br);
        st.nextToken();
        int cases = (int)st.nval;
		br.readLine();
		boolean firstCase = true;
		StringBuilder output = new StringBuilder();
		while(cases-- > 0) {
		    if(!firstCase)
		        output.append("\n");
		    ArrayList<ScoreBoard> scoreBoards = new ArrayList<>();
		    String s;
		    while((s = br.readLine()) != null && !s.isEmpty()) {
		        String[] tokens = s.split("\\s+");
		        short contestant = Short.parseShort(tokens[0]);
		        short problem = Short.parseShort(tokens[1]);
		        short time = Short.parseShort(tokens[2]);
		        char L = tokens[3].charAt(0);
		        int contestantPos = containContestant(scoreBoards, contestant);
		        if(contestantPos != -1) {
		            if(L == 'C') {
		                if(!scoreBoards.get(contestantPos).solvedProblems.contains(problem)) {
		                    scoreBoards.get(contestantPos).solvedProblems.add(problem);
		                    int problemPos = containProblem(scoreBoards.get(contestantPos).problemTried, problem);
		                    if(problemPos != -1) {
		                        scoreBoards.get(contestantPos).penaltyTime += time + 20 * scoreBoards.get(contestantPos).problemTried.get(problemPos).times;
		                        scoreBoards.get(contestantPos).problemTried.remove(problemPos);
		                    } else
		                        scoreBoards.get(contestantPos).penaltyTime += time;
		                }
		            } else if(L == 'I') {
		                int problemPos = containProblem(scoreBoards.get(contestantPos).problemTried, problem);
		                if(problemPos != -1)
		                    ++scoreBoards.get(contestantPos).problemTried.get(problemPos).times;
		                else
		                    scoreBoards.get(contestantPos).problemTried.add(new ProblemTried(problem, (short)1));
		            }
		        } else {
		            scoreBoards.add(new ScoreBoard(contestant));
		            if(L == 'C') {
		                scoreBoards.get(scoreBoards.size() - 1).solvedProblems.add(problem);
		                scoreBoards.get(scoreBoards.size() - 1).penaltyTime = time;
		            } else if(L == 'I')
		                scoreBoards.get(scoreBoards.size() - 1).problemTried.add(new ProblemTried(problem, (short)1));
		        }
		    }
		    Collections.sort(scoreBoards, new Comparator<ScoreBoard>() {
		        @Override
		        public int compare(ScoreBoard s1, ScoreBoard s2) {
		            if(s1.solvedProblems.size() != s2.solvedProblems.size())
		                return Integer.compare(s2.solvedProblems.size(), s1.solvedProblems.size());
		            if(s1.penaltyTime != s2.penaltyTime)
		                return Short.compare(s1.penaltyTime, s2.penaltyTime);
		            return Short.compare(s1.contestant, s2.contestant);
		        }
		    });
		    for(ScoreBoard S : scoreBoards)
		        output.append(S.contestant + " " + S.solvedProblems.size() + " " + S.penaltyTime + "\n");
		    firstCase = false;
		}
		System.out.print(output);
	}
	
	static int containContestant(ArrayList<ScoreBoard> arrayList, short key) {
	    for(int i = 0; i < arrayList.size(); ++i) {
	        if(arrayList.get(i).contestant == key)
	            return i;
	    }
	    return -1;
	}
	
	static int containProblem(ArrayList<ProblemTried> arrayList, short key) {
        for(int i = 0; i < arrayList.size(); ++i) {
            if(arrayList.get(i).problem == key)
                return i;
        }
        return -1;
    }
}

class ProblemTried {
    short problem, times;
    ProblemTried(short problem, short times) {
        this.problem = problem;
        this.times = times;
    }
}

class ScoreBoard {
    ArrayList<Short> solvedProblems;
    ArrayList<ProblemTried> problemTried;
    short contestant, penaltyTime;
    ScoreBoard(short contestant) {
        this.contestant = contestant;
        this.penaltyTime = 0;
        solvedProblems = new ArrayList<>();
        problemTried = new ArrayList<>();
    }
}

File: 10258 Contest Scoreboard/solution.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdbool.h>

typedef struct {
    short problem, times;
} ProblemTried;

typedef struct {
    short contestant, *solvedProblems, solvedCount, unsolvedCount, penaltyTime;
    ProblemTried *problemTried;
} ScoreBoard;

int containContestant(ScoreBoard *array, int size, short key) {
    for(int i = 0; i < size; ++i) {
        if(array[i].contestant == key)
            return i;
    }
    return -1;
}

bool containSolved(short *array, int size, short key) {
    for(int i = 0; i < size; ++i) {
        if(array[i] == key)
            return true;
    }
    return false;
}

int containUnsolved(ProblemTried *array, int size, short key) {
    for(int i = 0; i < size; ++i) {
        if(array[i].problem == key)
            return i;
    }
    return -1;
}

int compare(const void *a, const void *b) {
    ScoreBoard *s1 = (ScoreBoard*)a;
    ScoreBoard *s2 = (ScoreBoard*)b;
    if(s1->solvedCount != s2->solvedCount)
        return s1->solvedCount < s2->solvedCount;
    if(s1->penaltyTime != s2->penaltyTime)
        return s1->penaltyTime > s2->penaltyTime;
    return s1->contestant > s2->contestant;
}

int main() {
	int cases;
	scanf("%d  ", &cases);
    char *buffer = NULL;
	size_t bufsize = 0;
	while(cases--) {
	    ScoreBoard *scoreBoards = NULL;
	    int contestantCount = 0;
	    while(getline(&buffer, &bufsize, stdin) != -1 && buffer[0] != '\n') {
	        short contestant, problem, time;
	        char L;
	        sscanf(buffer, "%hd %hd %hd %c", &contestant, &problem, &time, &L);
	        int contestantPos = containContestant(scoreBoards, contestantCount, contestant);
    	    if(contestantPos != -1) {
    	        if(L == 'C') {
    	            if(!containSolved(scoreBoards[contestantPos].solvedProblems, scoreBoards[contestantPos].solvedCount, problem)) {
    	                scoreBoards[contestantPos].solvedProblems = (short*)realloc(scoreBoards[contestantPos].solvedProblems, (scoreBoards[contestantPos].solvedCount + 1) * sizeof(short));
        	            scoreBoards[contestantPos].solvedProblems[scoreBoards[contestantPos].solvedCount++] = problem;
    	                int problemPos = containUnsolved(scoreBoards[contestantPos].problemTried, scoreBoards[contestantPos].unsolvedCount, problem);
    	                if(problemPos != -1) {
            	            scoreBoards[contestantPos].penaltyTime += time + 20 * scoreBoards[contestantPos].problemTried[problemPos].times;
            	            scoreBoards[contestantPos].problemTried[problemPos] = scoreBoards[contestantPos].problemTried[--scoreBoards[contestantPos].unsolvedCount];
    	                } else
    	                    scoreBoards[contestantPos].penaltyTime += time;
    	            }
    	        } else if(L == 'I') {
    	            int problemPos = containUnsolved(scoreBoards[contestantPos].problemTried, scoreBoards[contestantPos].unsolvedCount, problem);
    	            if(problemPos != -1)
    	                ++scoreBoards[contestantPos].problemTried[problemPos].times;
    	            else {
    	                scoreBoards[contestantPos].problemTried = (ProblemTried*)realloc(scoreBoards[contestantPos].problemTried, (scoreBoards[contestantPos].unsolvedCount + 1) * sizeof(ProblemTried));
    	                scoreBoards[contestantPos].problemTried[scoreBoards[contestantPos].unsolvedCount].problem = problem;
    	                scoreBoards[contestantPos].problemTried[scoreBoards[contestantPos].unsolvedCount++].times = 1;
    	            }
    	        }
    	    } else {
    	        scoreBoards = (ScoreBoard*)realloc(scoreBoards, (contestantCount + 1) * sizeof(ScoreBoard));
    	        scoreBoards[contestantCount].contestant = contestant;
    	        scoreBoards[contestantCount].solvedProblems = NULL;
    	        scoreBoards[contestantCount].problemTried = NULL;
    	        scoreBoards[contestantCount].solvedCount = 0;
    	        scoreBoards[contestantCount].unsolvedCount = 0;
    	        scoreBoards[contestantCount].penaltyTime = 0;
    	        if(L == 'C') {
    	            scoreBoards[contestantCount].solvedProblems = (short*)realloc(scoreBoards[contestantCount].solvedProblems, (scoreBoards[contestantCount].solvedCount + 1) * sizeof(short));
    	            scoreBoards[contestantCount].solvedProblems[scoreBoards[contestantCount].solvedCount++] = problem;
    	            scoreBoards[contestantCount].penaltyTime = time;
    	        } else if(L == 'I') {
    	            scoreBoards[contestantCount].problemTried = (ProblemTried*)realloc(scoreBoards[contestantCount].problemTried, (scoreBoards[contestantCount].unsolvedCount + 1) * sizeof(ProblemTried));
    	            scoreBoards[contestantCount].problemTried[scoreBoards[contestantCount].unsolvedCount].problem = problem;
    	            scoreBoards[contestantCount].problemTried[scoreBoards[contestantCount].unsolvedCount++].times = 1;
    	        }
    	        ++contestantCount;
    	    }
	    }
	    qsort(scoreBoards, contestantCount, sizeof(ScoreBoard), compare);
	    for(int i = 0; i < contestantCount; ++i) {
	        printf("%hd %hd %hd\n", scoreBoards[i].contestant, scoreBoards[i].solvedCount, scoreBoards[i].penaltyTime);
	        free(scoreBoards[i].solvedProblems);
	        free(scoreBoards[i].problemTried);
	    }
	    free(scoreBoards);
	    if(cases)
	        putchar('\n');
	}
    free(buffer);
	return 0;
}
